caas/kubernetes/provider/provider.go

```diff
@@ -115,14 +123,34 @@ func (p kubernetesEnvironProvider) Open(args environs.OpenParams) (caas.Broker,
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	// Guinea Pig broker to hunt for the namespace where a controller lives. We
+	// disregard this one in favour of a new one pinned to the correct
+	// controller namespace when we find it.
 	broker, err := newK8sBroker(
-		args.ControllerUUID, k8sRestConfig, args.Config, newK8sClient,
-		newKubernetesNotifyWatcher, newKubernetesStringsWatcher, randomPrefix,
-		jujuclock.WallClock)
+		args.ControllerUUID, k8sRestConfig, args.Config, DefaultNamespace,
```

Comment:
Shouldn't we be constructing the broker with broker.Config().Name() here?
We return it on line 140 below with an incorrect namespace if model is not "controller".
In which case, the "Guinea Pig" comment is slightly inaccurate because what we are doing is constructing what we think is the broker to use, and only doing something else if it is a controller model.

I think the flow should be:

1. look at broker.Config().Name()
2. if controller model, construct generic broker to find ns to use
3. construct result broker to return with the correct namespace

Reply:
Take a look at that. Thanks for the pickup